provider/workspace_preset.go

```diff
@@ -0,0 +1,64 @@
+package provider
+
+import (
+	"context"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/mitchellh/mapstructure"
+)
+
+type WorkspacePreset struct {
+	Name       string            `mapstructure:"name"`
+	Parameters map[string]string `mapstructure:"parameters"`
+}
+
+func workspacePresetDataSource() *schema.Resource {
+	return &schema.Resource{
+		SchemaVersion: 1,
+
+		Description: "Use this data source to predefine common configurations for workspaces.",
+		ReadContext: func(ctx context.Context, rd *schema.ResourceData, i interface{}) diag.Diagnostics {
+			var preset WorkspacePreset
+			err := mapstructure.Decode(struct {
+				Name       interface{}
+				Parameters interface{}
+			}{
+				Name:       rd.Get("name"),
+				Parameters: rd.Get("parameters"),
+			}, &preset)
+			if err != nil {
+				return diag.Errorf("decode workspace preset: %s", err)
+			}
+
+			if preset.Name == "" {
+				return diag.Errorf("workspace preset name must be set")
+			}
+
+			if len(preset.Parameters) == 0 {
```

Comment:
This should be implemented as validation options in the schema.

See `MinItems` for `parameters`.

Interesting/weird that the validation doesn't catch empty strings with its `Required` attribute. Thanks for checking both 👍

Reply:
MinItems doesn't seem to work with Maps. Only with lists or sets. It's the one part of custom validation that I've needed to retain. I've been able to remove all the rest in favour of adding the the Schema though.